common/codex32.c

```diff
@@ -0,0 +1,170 @@
+
+	uint8_t next_byte = 0;
+	uint8_t rem = 0;
+	size_t i = 0, j = 0;
+	while (parts->payload[i] != '\0') {
+		char ch = parts->payload[i++];
+		uint8_t fe = bech32_charset_rev[(int)ch];
+
+		if (rem < 3) {
+			// If we are within 3 bits of the start we can fit the whole next char in
+			next_byte |= fe << (3 - rem);
+		}
+		else if (rem == 3) {
+			// If we are exactly 3 bits from the start then this char fills in the byte
+			ret[j++] = next_byte | fe;
+			next_byte = 0;
+		}
+		else { // rem > 3
+			// Otherwise we have to break it in two
+			uint8_t overshoot = rem - 3;
+			assert(overshoot > 0);
+			ret[j++] = next_byte | (fe >> overshoot);
+			next_byte = fe << (8 - overshoot);
+		}
+
+		rem = (rem + 5) % 8;
+	}
+	assert(rem <= 4); // checked when parsing the string
+	return ret;
+}
+
+static bool case_check(const char *codex32str)
+{
+	bool have_lower = false, have_upper = false;
+	size_t str_len = strlen(codex32str);
+	for (size_t i = 0; i < str_len; i++) {
+		if (codex32str[i] >= 'a' && codex32str[i] <= 'z') {
+            		have_lower = true;
+		} else if (codex32str[i] >= 'A' && codex32str[i] <= 'Z') {
+			have_upper = true;
+		}
+	}
+	if (have_lower && have_upper) {
+		return false;
+	}
+	return true;
+}
+
+/* Return `False` if the codex32 is invalid */
+bool codex32_decode(const char *codex32str,
+		    struct codex32_parts *parts,
+		    const u8 **payload,
+		    char **fail)
+{
+	size_t checksum_len;
+	char *sep = strchr(codex32str, '1');
+	size_t codex32str_len = strlen(codex32str);
+
+	// Separator `1` doesn't exist, Invalid codex string!
+	if (!sep) {
+		*fail = tal_fmt(tmpctx, "Separator doesn't exist!");
+		return false;
+	}
+
+	if (!case_check(codex32str)) {
+		*fail = tal_fmt(tmpctx, "Case inconsistency!");
+		return false;
+	}
+
+	const char *hrp = tal_strndup(parts, codex32str, sep - codex32str),
+	           *codex_datastr = tal_strndup(parts,
+		   			        codex32str + (sep - codex32str + 1),
+					        strlen(codex32str) - strlen(hrp) + 1);
+
+	if (!(streq(hrp, "ms") || streq(hrp, "MS"))) {
+		*fail = tal_fmt(tmpctx, "Invalid HRP!");
+		return false;
+	}
+
+	/* FIXME: Confirm if the numbers are correct. */
+	if (codex32str_len >= 48 && codex32str_len < 94) {
+		parts->codexl = 0;
+	} else if (codex32str_len >= 125 && codex32str_len < 128) {
+		parts->codexl = 1;
+	} else {
+		*fail = tal_fmt(tmpctx, "Invalid length!");
+		return false;
+	}
+
+	if (strlen(codex_datastr) > 93) {
+		checksum_len = 15;
+	} else {
+		checksum_len = 13;
+	}
+
+	parts->hrp = hrp;
+	parts->threshold = codex_datastr[0] - '0';;
+	parts->id = tal_strndup(parts, codex_datastr + 1, 4);
+	parts->share_idx = tal_strndup(parts, codex_datastr + 5, 1);
+	parts->payload = tal_strndup(parts,
+				     codex_datastr + 6,
+				     strlen(codex_datastr) - checksum_len - 6);
+	parts->checksum = tal_strndup(parts,
+				      codex_datastr + strlen(parts->payload) + 6,
+				      checksum_len);
```

Comment:
This risks overrun.  I know you manually check length above, but I think these helpers would be better:

```
static const char *pull_front_bytes(const tal_t *ctx, const char **cursor, size_t len)
{
    const char *ret;
     if (strlen(*cursor) < len))
           return NULL;
    ret = tal_strndup(ctx, *cursor, len);
    *cursor += len;
    return ret;
}

/* Pull all bytes except for `leave` at the end */
static const char *pull_remaining_bytes(const tal_t *ctx, const char **cursor, size_t leave)
{
    size_t slen = strlen(*cursor);
   if (slen < leave)
        return NULL;
    slen -= leave;
    return pull_front_bytes(ctx, cursor, slen);
}
```